### PR TITLE
TLS Tests

### PR DESCRIPTION
**Pull request by glbrntt**
- Initial tests only include variants for the "normal" functional tests
- Adds an additional target with self signed certificates; by avoiding the filesystem we also avoid having to run the Echo demo from the right directory
- Tidy up test helpers

### CODE REVIEW

File: Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import NIOSSL
+
+/// Wraps `NIOSSLCertificate` to provide the certificate common name and expiry date.
+public struct GRPCSwiftCertificate {
```
Comment (MrMage) [inline Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift]: nit: is there a way to make it even clearer naming-wise that these are just intended as sample certificates? E.g. `SampleCertificate`? Just an idea.

Similarly, the `TestData` module name feels a bit weird to me. How about `SwiftGRPCNIO{Ex,S}ampleData`? Still not yet sure myself how much better these would be, let's discuss this.

File: Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  /// Returns whether the certificate has expired.
+  public var isExpired: Bool {
+    return notAfter < Date()
+  }
+}
+
+/// Provides convenience methods to make `NIOSSLPrivateKey`s for corresponding `GRPCSwiftCertificate`s.
+public struct GRPCSwiftPrivateKey {
+  private init() {
+  }
```
Comment (MrMage) [inline Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift]: super-nit: move to the above line?

File: Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  public static func server() throws -> NIOSSLPrivateKey {
+    return try NIOSSLPrivateKey(buffer: Array(serverKey.utf8CString), format: .pem)
+  }
+
+  public static func client() throws -> NIOSSLPrivateKey {
+    return try NIOSSLPrivateKey(buffer: Array(clientKey.utf8CString), format: .pem)
+  }
+}
+
+// MARK:- Certificates and private keys
```
Comment (MrMage) [inline Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift]: super nit: add a space after the colon?

File: Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+import Foundation
+import NIOSSL
+
+/// Wraps `NIOSSLCertificate` to provide the certificate common name and expiry date.
+public struct GRPCSwiftCertificate {
+  public var certificate: NIOSSLCertificate
+  public var commonName: String
+  public var notAfter: Date
+
+  public static func ca() throws -> GRPCSwiftCertificate {
```
Comment (MrMage) [inline Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift]: make these static properties that only get initialized once instead?

File: Tests/SwiftGRPCNIOTests/NIOBasicEchoTestCase.swift [hunk 5]
```diff
@@ -31,41 +33,132 @@ extension Echo_EchoResponse {
...
 
-class NIOBasicEchoTestCase: XCTestCase {
+enum TransportSecurity {
+  case none
+  case anonymousClient
+  case mutualAuthentication
+}
+
+extension TransportSecurity {
+  private func caCert() throws -> GRPCSwiftCertificate {
```
Comment (MrMage) [inline Tests/SwiftGRPCNIOTests/NIOBasicEchoTestCase.swift]: the _could_ also be static properties I guess

File: Tests/SwiftGRPCNIOTests/NIOBasicEchoTestCase.swift [hunk 6]
```diff
@@ -31,41 +33,132 @@ extension Echo_EchoResponse {
...
+  }
+}
+
+class NIOEchoTestCaseBase: XCTestCase {
   var defaultTestTimeout: TimeInterval = 1.0
 
-  var serverEventLoopGroup: EventLoopGroup!
-  var server: GRPCServer!
+  var serverEventLoopGroup: EventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+  var clientEventLoopGroup: EventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
```
Comment (MrMage) [inline Tests/SwiftGRPCNIOTests/NIOBasicEchoTestCase.swift]: `let`? (Also below)

File: Tests/SwiftGRPCNIOTests/NIOFunctionalTests.swift [hunk 7]
```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2018, gRPC Authors All rights reserved.
```
Comment (MrMage) [inline Tests/SwiftGRPCNIOTests/NIOFunctionalTests.swift]: super-nit: 2019? (also elsewhere)

File: Tests/SwiftGRPCNIOTests/NIOFunctionalTests.swift [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+ */
+import Dispatch
+import Foundation
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+@testable import SwiftGRPCNIO
+import XCTest
+
+protocol NIOFunctionalTests: class {
```
Comment (MrMage) [inline Tests/SwiftGRPCNIOTests/NIOFunctionalTests.swift]: What's the advantage of putting these into a protocol?

File: Tests/SwiftGRPCNIOTests/EventLoopFuture+Assertions.swift [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import XCTest
+
+extension EventLoopFuture where Value: Equatable {
+  /// Registers a callback which asserts the value promised by this future is equal to
+  /// the expected value. Causes a test failure if the future returns an error.
+  ///
+  /// - Parameters:
+  ///   - expected: The expected value.
+  ///   - expectation: A test expectation to fulfill once the future has completed.
+  func assertEqual(_ expected: Value, fulfill expectation: XCTestExpectation? = nil, file: StaticString = #file, line: UInt = #line) {
```
Comment (MrMage) [inline Tests/SwiftGRPCNIOTests/EventLoopFuture+Assertions.swift]: is there a reasonable case where `expectation` would be nil? In that case we could never be sure that our check ever runs, I think.

File: Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift (hunk 1 above)
Comment (glbrntt) [reply Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift]: `SampleCertificate` is better. `SwiftGRPCNIOSampleData` works for me too, I wasn't keen on the naming before!

File: Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift (hunk 4 above)
Comment (glbrntt) [reply Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift]: We'd have to `try!` if they were static properties; I _guess_ this isn't so much an issue as we know it works with the current certificates. What do you think?

File: Tests/SwiftGRPCNIOTests/NIOFunctionalTests.swift (hunk 8 above)
Comment (glbrntt) [reply Tests/SwiftGRPCNIOTests/NIOFunctionalTests.swift]: We can use `Self` in the default implementation:

```swift
static var allTests: [(String, (Self) -> () throws -> Void)]
```

If we didn't then I think we'd have to override this in each subclass.

I was also thinking we should use `swift test --generate-linuxmain` or similar script instead but wanted to save that for a different PR.

File: Tests/SwiftGRPCNIOTests/EventLoopFuture+Assertions.swift (hunk 9 above)
Comment (glbrntt) [reply Tests/SwiftGRPCNIOTests/EventLoopFuture+Assertions.swift]: Good point, there probably isn't a case. If we find one we can easily add this back.

**Comment by glbrntt** [comment]
> Initial impression is very good. I must admit I haven't yet through each test in `NIOFunctionalTests.swift`, though.

They're more or less the `NIOServerTests` but relying on `XCTestExpectation`s instead of call timeouts instead.

File: Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift (hunk 4 above)
Comment (MrMage) [reply Sources/SwiftGRPCNIOTestData/GRPCSwiftCertificate.swift]: I think the `try!` should be fine here.

File: Tests/SwiftGRPCNIOTests/NIOFunctionalTests.swift (hunk 8 above)
Comment (MrMage) [reply Tests/SwiftGRPCNIOTests/NIOFunctionalTests.swift]: Ah, thanks for the elaboration. Let's stick with the protocol in this PR and run `swift test --generate-linuxmain` in a follow-up PR (possibly adding it to the `Makefile`; we could even check that this doesn't update anything as part of CI).

**Pull request merged**